### PR TITLE
Avoid retry when server error

### PR DESCRIPTION
**Pull request by antollinim**

### CODE REVIEW

File: extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java [hunk 1]
```diff
@@ -338,7 +336,8 @@ private boolean shouldRetry(HttpRequestBase method, Throwable t, int retries) {
 
         if (t instanceof NoHttpResponseException ||
                 t instanceof SocketException ||
-                t instanceof SocketTimeoutException) {
+                t instanceof SocketTimeoutException ||
+                t instanceof ConnectTimeoutException) {
```
Comment (antollinim) [inline extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java]: @lhazlewood We were not previously retrying on ConnectionTimeout. Was it unintentionally left out? I added that feature now.  Please let me know if I should keep the previous behavior (i.e. not retrying on ConnectionTimeout). Thanks!

File: extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java [hunk 2]
```diff
@@ -246,10 +247,7 @@ public Response executeRequest(Request request) throws RestException {
                         throw new RestException("HTTP 429: Too Many Requests.  Exceeded request rate limit in the allotted amount of time.");
                     }
 
-                    if (!response.isServerError() || retryCount > this.numRetries) {
-                        return response;
-                    }
-                    //otherwise allow the loop to continue to execute a retry request
```
Comment (antollinim) [inline extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java]: Retries do not happen in the code block any longer. They take place on `Exceptions`, not on proper server error responses.

File: extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java (hunk 1 above)
Comment (lhazlewood) [inline extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java]: We should retry on ConnectionTimeout.

File: extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java (hunk 2 above)
Comment (lhazlewood) [inline extensions/httpclient/src/main/java/com/stormpath/sdk/impl/http/httpclient/HttpClientRequestExecutor.java]: We *should* retry on server error - if it is a 503 for example, we should definitely retry.

**Comment by lhazlewood** [comment]
Why was this issue created? What is wrong with retrying on server error?

If it is a 503 for example, there should absolutely be a retry because the server will likely be available soon enough.

**Comment by antollinim** [comment]
The issue was created because a few days ago there was a bug in the Backend (500 error) and the backend was getting some spikes because of the error+continual retries.

I agree that not all the server errors are the same, I just did it to broad, my bad. I should only avoid retrying on 500 specifically. Do you agree?

**Comment by lhazlewood** [comment]
I looked at the status codes we are likely to return and we should retry on 503 or 504 - all others should not be retried.

**Comment by antollinim** [comment]
@lhazlewood Thanks for the clarification. Code has been updated accordingly.

**Comment by lhazlewood** [comment]
Let's get the build to succeed and I'll merge.

**Pull request merged**